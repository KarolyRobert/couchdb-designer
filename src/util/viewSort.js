
const byNumber = (a,b) => {
    return a-b
}

const byString = (a,b) => {
    if(a === b){
        return 0;
    }
    if(a > b){
        return 1;
    }else{
        return -1;
    }
}
const byObject = (a,b) => {
    return byString(JSON.stringify(a),JSON.stringify(b))
}

const byArray = (a,b) => {
    let index = 0;
    let keyType = Array.isArray(a[index]) ? 'array' : typeof a[index];
    while(joker[keyType](a[index],b[index]) === 0){
        if(a.length > index+1 && b.length > index+1){
            index++;
            keyType = Array.isArray(a[index]) ? 'array' : typeof a[index];
        }else{
            return a.length - b.length;
        }
    }
    return joker[keyType](a[index],b[index]);
}

const joker = {
    number:byNumber,
    string:byString,
    object:byObject,
    array:byArray
}

const viewSort = {
    number:(a,b) => byNumber(a.key,b.key),
    string:(a,b) => byString(a.key,b.key),
    object:(a,b) => byObject(a.key,b.key),
    array:(a,b) => byArray(a.key,b.key)
}

export default viewSort;