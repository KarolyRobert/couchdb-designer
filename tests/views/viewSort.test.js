import viewSort from '../../src/testing/views/viewSort';



describe('viewSort',() => {
    test('sort',() => {
        let rows = [
            {key:{b:10,a:1}},
            {key:{b:10}},
            {key:{b:10}},
            {key:{a:false, b:true}},
            {key:{a:10}},
            {key:{a:null}},
            {key:[1,false,true]},
            {key:[1,null,23]},
            {key:[0,20,30]},
            {key:[null,0,'zebra']},
            {key:[0,20,10]},
            {key:'zebra'},
            {key:'alma'},
            {key:10},
            {key:0},
            {key:true},
            {key:false},
            {key:null}
        ]
        rows.sort(viewSort);
        expect(rows).toMatchSnapshot();
    })
})