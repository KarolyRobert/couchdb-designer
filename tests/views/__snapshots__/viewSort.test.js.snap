// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewSort sort 1`] = `
Array [
  Object {
    "key": null,
  },
  Object {
    "key": false,
  },
  Object {
    "key": true,
  },
  Object {
    "key": 0,
  },
  Object {
    "key": 10,
  },
  Object {
    "key": "alma",
  },
  Object {
    "key": "zebra",
  },
  Object {
    "key": Array [
      null,
      0,
      "zebra",
    ],
  },
  Object {
    "key": Array [
      0,
      20,
      10,
    ],
  },
  Object {
    "key": Array [
      0,
      20,
      30,
    ],
  },
  Object {
    "key": Array [
      1,
      null,
      23,
    ],
  },
  Object {
    "key": Array [
      1,
      false,
      true,
    ],
  },
  Object {
    "key": Object {
      "a": null,
    },
  },
  Object {
    "key": Object {
      "a": false,
      "b": true,
    },
  },
  Object {
    "key": Object {
      "a": 10,
    },
  },
  Object {
    "key": Object {
      "b": 10,
    },
  },
  Object {
    "key": Object {
      "b": 10,
    },
  },
  Object {
    "key": Object {
      "a": 1,
      "b": 10,
    },
  },
]
`;
